charmrepo/charmstore.go

```diff
@@ -63,11 +69,65 @@ func NewCharmStore(p NewCharmStoreParams) (Interface, error) {
 	}, nil
 }
 
-var notImplemented = errgo.New("not implemented")
-
 // Get implements Interface.Get.
 func (s *CharmStore) Get(curl *charm.URL) (charm.Charm, error) {
-	return nil, notImplemented
+	// Disable bundle deployment.
+	if curl.Series == "bundle" {
+		return nil, errgo.Newf("expected a charm URL, got bundle URL %q", curl)
+	}
+
+	// Prepare the cache directory and retrieve the charm.
+	if err := os.MkdirAll(s.cacheDir, os.FileMode(0755)); err != nil {
+		return nil, errgo.Notef(err, "cannot create the cache directory")
+	}
+	r, id, expectHash, _, err := s.client.GetArchive(curl.Reference())
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot retrieve charm")
+	}
+	defer r.Close()
+
+	// Check if the archive already exists in the cache.
+	path := filepath.Join(s.cacheDir, charm.Quote(id.String())+".charm")
+	if verifyHash384(path, expectHash) == nil {
```

Comment:
I wonder if we should be using sha256 for this so that we don't blow everyone's cache.